### PR TITLE
Refactoring and pulling state up

### PR DESCRIPTION
**Pull request by tanishqkancharla**
Refactoring and pulling state up to the root, in preparation for pulling it out of `useState`. 

This didn’t change any behavior except that instead of adding an empty card when its not an error or when its already submitted,
```ts
if (!isError && !alreadySubmitted) addEmptyCard();
```

it adds a new card only when its not the last one. This is so you don’t get “stuck” on the last pane when you run into an error.
```ts
if (!isLast) addEmptyCard();
```

### CODE REVIEW

**Comment by hustcer** [comment]
Thanks, and why should we use `use-state-dispatch` ? It may not be very mature yet

**Comment by tanishqkancharla** [comment]
I wrote the library 😅 its a very tiny helper hook I use in a lot of projects, like 75 lines of code.

**Comment by rgwood** [comment]
Thanks for the PR!

I'm also reluctant to add `use-state-dispatch`. Nothing against your library, but we're light on experience with web development and we don't have the expertise or time to evaluate it.

What's your thinking behind adding [Nanoid](https://www.npmjs.com/package/nanoid) for card IDs? I'm not sure I see the advantage compared to an integer ID.

**Comment by tanishqkancharla** [comment]
Ok, no worries, I can just use useReducer instead. Nanoid, because I believe the previous way had bugs (if you deleted a card, the next card created could have a duplicate id). I just picked nanoid as a uuid library I’m familiar with. Tbh I can also just replace that with `Math.random().slice(0, 13).toString()` or something

**Comment by rgwood** [comment]
Ahh, that makes sense; Nanoid's fine by me, that sounds like a decent solution to the problem.

**Comment by tanishqkancharla** [comment]
Removed both `use-state-dispatch` and `nanoid`. No need to use `nanoid` over the random string if we don’t need to.

**Comment by rgwood** [comment]
Potentially a dumb question, but: in what sense is this PR pulling state up to the root? I can see that the PR is changing `useState` to `useReducer` but it doesn't seem like the location of state has changed.

It looks good overall but I feel like I might be missing something.

File: src/app/Card.tsx [hunk 1]
```diff
@@ -1,33 +1,38 @@
...
-  history,
-  output,
-  onClose,
-  onSubmit,
-}: ICard & {
-  history: string[];
-  onSubmit: (newProps: Partial<CardPropTypes>, isError: boolean) => void;
-  onClose: () => void;
-}) => {
-  const [input, setInput] = useState(initialInput);
```
Comment (tanishqkancharla) [inline src/app/Card.tsx:30]: I pulled this piece of state up. Now it just belongs to the card, and gets updated via `onInputChange` in App. The `activeHistoryIndex` state still belongs to this component, that hasn’t been pulled up yet. I’m not sure it needs to, though.

**Comment by tanishqkancharla** [comment]
https://github.com/nushell/nana/pull/48/files#r905680876

File: src/app/Card.tsx (hunk 1 above)
Comment (rgwood) [reply src/app/Card.tsx:30]: Makes sense, thanks

File: src/app/App.tsx [hunk 2]
```diff
@@ -1,59 +1,85 @@
...
-  };
+      });
+  }
+};
+
+export const App = () => {
+  const [history, dispatchHistory] = useReducer(historyReducer, []);
+  const [cards, dispatchCards] = useReducer(cardsReducer, []);
+
+  const addEmptyCard = useCallback(async () => {
```
Comment (rgwood) [inline src/app/App.tsx:58]: My understanding is that [`useCallBack`](https://reactjs.org/docs/hooks-reference.html#usecallback) is a performance optimization; is that correct?

I don't know enough about React to understand why the optimization is important/worthwhile here.

Comment (tanishqkancharla) [reply src/app/App.tsx:58]: That’s right. In this case, I think its not really necessary. Did it out of habit. I’ll remove.

Comment (tanishqkancharla) [reply src/app/App.tsx:58]: Honestly in this case, I think its worse to have it.

File: src/app/Card.tsx [hunk 3]
```diff
@@ -1,33 +1,38 @@
...
+export const Card = (
+  props: ICard & {
+    history: string[];
+    onInputChange: (newInput: string) => void;
+    onSubmit: (newProps: CardPropTypes, isError: boolean) => void;
+    onClose: () => void;
+  }
+) => {
```
Comment (rgwood) [inline src/app/Card.tsx:25]: I find this easier to read than the old destructuring-inside-parens approach, thanks.

File: yarn.lock [hunk 4]
```diff
@@ -4,27 +4,27 @@
 
 "@ampproject/remapping@^2.1.0":
   version "2.2.0"
-  resolved "https://registry.yarnpkg.com/@ampproject/remapping/-/remapping-2.2.0.tgz#56c133824780de3174aed5ab6834f3026790154d"
+  resolved "https://registry.npmjs.org/@ampproject/remapping/-/remapping-2.2.0.tgz"
```
Comment (rgwood) [inline yarn.lock:7]: Almost done! Last question, I think: are these lockfile changes intentional? Looks like something that might have been done by accident

Comment (tanishqkancharla) [reply yarn.lock:7]: Oof no. I accidentlally did npm install at the beginning. I thought removing package-lock and doing yarn install would essentially reverse what I did…

Comment (tanishqkancharla) [reply yarn.lock]: Idk. I tried removing yarn lock, and redoing yarn install, and there’s still a diff…

**Comment by rgwood** [comment]
LGTM, thanks for the contribution!

**Pull request merged**